// flatted/index.js
self.Flatted = (function (exports) {
  'use strict';

  

  /*! (c) 2020 Andrea Giammarchi */
  let $parse = JSON.parse,
      $stringify = JSON.stringify;
  let keys = Object.keys;
  let Primitive = String; // it could be Number

  let primitive = 'string'; // it could be 'number'

  let ignore = {};
  let object = 'object';

  let noop = function noop(_, value) {
    return value;
  };

  let primitives = function primitives(value) {
    return value instanceof Primitive ? Primitive(value) : value;
  };

  let Primitives = function Primitives(_, value) {
    return typeof(value) === primitive ? new Primitive(value) : value;
  };

  let revive = function revive(input, parsed, output, $) {
    let lazy = [];

    for (let ke = keys(output), length = ke.length, y = 0; y < length; y++) {
      let k = ke[y];
      let value = output[k];

      if (value instanceof Primitive) {
        let tmp = input[value];

        if (typeof(tmp) === object && !parsed.has(tmp)) {
          parsed.add(tmp);
          output[k] = ignore;
          lazy.push({
            k: k,
            a: [input, parsed, tmp, $]
          });
        } else output[k] = $.call(output, k, tmp);
      } else if (output[k] !== ignore) output[k] = $.call(output, k, value);
    }

    for (let _length = lazy.length, i = 0; i < _length; i++) {
      let _lazy$i = lazy[i],
          _k = _lazy$i.k,
          a = _lazy$i.a;
      output[_k] = $.call(output, _k, revive.apply(null, a));
    }

    return output;
  };

  let set = function set(known, input, value) {
    let index = Primitive(input.push(value) - 1);
    known.set(value, index);
    return index;
  };

  let parse = function parse(text, reviver) {
    let input = $parse(text, Primitives).map(primitives);
    let value = input[0];
    let $ = reviver || noop;
    let tmp = typeof(value) === object && value ? revive(input, new Set(), value, $) : value;
    return $.call({
      '': tmp
    }, '', tmp);
  };
  let stringify = function stringify(value, replacer, space) {
    let $ = replacer && typeof(replacer) === object ? function (k, v) {
      return k === '' || -1 < replacer.indexOf(k) ? v : void 0;
    } : replacer || noop;
    let known = new Map();
    let input = [];
    let output = [];
    let i = +set(known, input, $.call({
      '': value
    }, '', value));
    let firstRun = !i;

    while (i < input.length) {
      firstRun = true;
      output[i] = $stringify(input[i++], replace, space);
    }

    return '[' + output.join(',') + ']';

    function replace(key, value) {
      if (firstRun) {
        firstRun = !firstRun;
        return value;
      }

      let after = $.call(this, key, value);

      switch (typeof(after)) {
        case object:
          if (after === null) return after;

        case primitive:
          return known.get(after) || set(known, input, after);
      }

      return after;
    }
  };
  let toJSON = function toJSON(any) {
    return $parse(stringify(any));
  };
  let fromJSON = function fromJSON(any) {
    return parse($stringify(any));
  };

  exports.fromJSON = fromJSON;
  exports.parse = parse;
  exports.stringify = stringify;
  exports.toJSON = toJSON;

  return exports;

})({});
